### PR TITLE
Add support for linking against fat/universal dynamic libraries on macOS

### LINKED ISSUE #9169: zig ld: ability to link against universal binaries
Problem: Some libraries on macOS are distributed as universal binaries that might be the only option to link against - even when building for a single architecture.

Solution: Support linking against universal binaries

Universal binaries on macOS work by embedding the data for multiple libraries (one for each arch)  into a single file - with a "fat" header to provide the offsets for each of these. 

The format of this header is quite simple (a magic number along with the number of headers, followed by a list of headers containing the appropriate offsets). The difficulty here is making sure that zld doesn't assume one file == one library, and that we can then find the correct offset for any libraries we might want to link against for a given target. 

See the discussion in #8317 for more context.

### PR DESCRIPTION
**Pull request by moosichu**
This resolves #9169.

I've validated that this works by successfully linking with the vulkan dylibs on macOS, which are shipped as fat aarch64+x86_64 libraries.

### CODE REVIEW

File: src/link/MachO/Dylib.zig
Comment (LemonBoy) [inline src/link/MachO/Dylib.zig:78]: Given `zld` assumes all the input mach-o files are 64bit ones I'd place the fat-parsing code here (or, better, in a subroutine that's called here).
The base offset can be computed once, according to the selected architecture, and then stored in `Dylib` to avoid passing `seek_pos` everywhere.

**Review by LemonBoy — COMMENTED** [review]
Small drive-by code review, I hope @kubkon doesn't mind.

File: src/link/MachO/Dylib.zig
Comment (kubkon) [reply src/link/MachO/Dylib.zig:193]: Adding a check for `MH_MAGIC_64` or `MH_CIGAM_64` would be nice, and it should also be propagated to `Object.isObject`. The ultimate purpose of `isDylib` and `isObject` is to work out which parser to use `Object.parse` or `Dylib.parse` in zld, nothing more.

File: lib/std/io/reader.zig
Comment (daurnimator) [reply lib/std/io/reader.zig:309]: I think this has come up before.

A quick search around just found https://github.com/ziglang/zig/pull/8108#discussion_r584686657 though

File: src/link/MachO/Dylib.zig
Comment (kubkon) [inline src/link/MachO/Dylib.zig:205]: As implied by @LemonBoy, we should leave the decision making whether this FAT dylib contains the architecture of interest to the actual parsing logic, and then warn the user if it doesn't but don't trigger a hard error. I guess a hard error should only ever occur if the parsing cannot be performed, read, the dylib is malformed.

File: src/link/MachO/Dylib.zig
Comment (kubkon) [reply src/link/MachO/Dylib.zig:78]: It might be worth looking at how this is done in `Archive.zig` which after all is a collection of objects, so naturally we need to seek to find the right object.

Comment (kubkon) [reply src/link/MachO/Dylib.zig:78]: Oh and we should not have to parse *all* dylibs within a fat dylib, only the one matching the requested architecture.

File: src/link/MachO/Zld.zig
Comment (kubkon) [inline src/link/MachO/Zld.zig:1837]: Proxies should never collide meaning `self.imports.putNoClobber(...)` should never trap. If that is the case, the linking will most likely result in missing links between `Undef`s and `Proxy`s.

Comment (kubkon) [reply src/link/MachO/Zld.zig:1837]: Either way, this logic is completely rewritten as part of #9171 so you might wanna wait for this to land before tweaking the symbol resolution logic.

**Comment by kubkon** [comment]
> Small drive-by code review, I hope @kubkon doesn't mind.

Not at all!

**Comment by kubkon** [comment]
Just something to think about: I thought that a universal binary is just a collection of dylibs per architecture that are well-formed (hence fat since they explode in size I guess). With this in mind, I wouldn't expect any need to tweak any symbol resolution logic; rather, I'd expect a helper function which takes a file descriptor to a fat dylib and required architecture, and based on this, creates a `*Dylib` struct. zld would then simply append this `*Dylib` to `self.dylibs` `ArrayList` and that should be it. Does that make sense?

File: src/link/MachO/Zld.zig
Comment (kubkon) [reply src/link/MachO/Zld.zig:1837]: See the comment below: I'm surprised you need to touch the symbol resolution logic at all.

**Comment by moosichu** [comment]
Cheers! Thanks for the feedback! Yeah - thanks for reviewing the obviously bad stuff & giving feedback on how you would resolve that within the context of the compiler. Will take it all on-board :)

**Comment by kubkon** [comment]
> Cheers! Thanks for the feedback! Yeah - thanks for reviewing the obviously bad stuff & giving feedback on how you would resolve that within the context of the compiler. Will take it all on-board :)

Thanks for working on this @moosichu! Feel free to ping me here or on Discord if you get stuck or need to discuss something - I'll be happy to unblock you!

File: src/link/MachO/Zld.zig
Comment (moosichu) [reply src/link/MachO/Zld.zig:1837]: Yeah - it's worth pointing out that this was just a short-term hack to just get an example working. I'm getting some kind of conflict with the library I've been testing with - I just don't know why & this was a very hacky way to just work around that.

File: src/link/MachO/Dylib.zig
Comment (moosichu) [reply src/link/MachO/Dylib.zig:205]: At the moment if you try to link with a dylib that doens't match the requested architecture you _do_ get a hard error - so I'm going to match that existing logic for now, but can change both to a warning if you think that would be best afterwards.

File: src/link/MachO/Zld.zig
Comment (moosichu) [reply src/link/MachO/Zld.zig:1837]: So the reason I added that log warnings was to see why that was happening - and every proxy.name appears twice (back-to-back) for some reason. This is what i get when linking with the vulkan dynamic libraries:

```
warning(zld): _vkQueuePresentKHR
warning(zld): _vkQueuePresentKHR
warning(zld): _vkGetSwapchainImagesKHR
warning(zld): _vkGetSwapchainImagesKHR
warning(zld): _vkGetPhysicalDeviceSurfaceSupportKHR
warning(zld): _vkGetPhysicalDeviceSurfaceSupportKHR
warning(zld): _vkGetPhysicalDeviceSurfaceFormatsKHR
warning(zld): _vkGetPhysicalDeviceSurfaceFormatsKHR
warning(zld): _vkGetPhysicalDeviceSurfaceCapabilitiesKHR
warning(zld): _vkGetPhysicalDeviceSurfaceCapabilitiesKHR
warning(zld): _vkGetPhysicalDeviceQueueFamilyProperties
warning(zld): _vkGetPhysicalDeviceQueueFamilyProperties
warning(zld): _vkGetPhysicalDeviceMemoryProperties
warning(zld): _vkGetPhysicalDeviceMemoryProperties
warning(zld): _vkGetDeviceQueue
warning(zld): _vkGetDeviceQueue
warning(zld): _vkFreeMemory
warning(zld): _vkFreeMemory
warning(zld): _vkEnumeratePhysicalDevices
warning(zld): _vkEnumeratePhysicalDevices
warning(zld): _vkEndCommandBuffer
warning(zld): _vkEndCommandBuffer
warning(zld): _vkDestroySwapchainKHR
warning(zld): _vkDestroySwapchainKHR
warning(zld): _vkDestroyInstance
warning(zld): _vkDestroyInstance
warning(zld): _vkDestroyDevice
warning(zld): _vkDestroyDevice
warning(zld): _vkDestroyDescriptorPool
warning(zld): _vkDestroyDescriptorPool
warning(zld): _vkDestroyCommandPool
warning(zld): _vkDestroyCommandPool
warning(zld): _vkCreateSwapchainKHR
warning(zld): _vkCreateSwapchainKHR
warning(zld): _vkCreateShaderModule
warning(zld): _vkCreateShaderModule
warning(zld): _vkCreateSemaphore
warning(zld): _vkCreateSemaphore
warning(zld): _vkCreatePipelineLayout
warning(zld): _vkCreatePipelineLayout
warning(zld): _vkCreatePipelineCache
warning(zld): _vkCreatePipelineCache
warning(zld): _vkCreateInstance
warning(zld): _vkCreateInstance
warning(zld): _vkCreateImageView
warning(zld): _vkCreateImageView
warning(zld): _vkCreateDevice
warning(zld): _vkCreateDevice
warning(zld): _vkCreateDescriptorSetLayout
warning(zld): _vkCreateDescriptorSetLayout
warning(zld): _vkCreateDescriptorPool
warning(zld): _vkCreateDescriptorPool
warning(zld): _vkCreateComputePipelines
warning(zld): _vkCreateComputePipelines
warning(zld): _vkCreateCommandPool
warning(zld): _vkCreateCommandPool
warning(zld): _vkCmdBindPipeline
warning(zld): _vkCmdBindPipeline
warning(zld): _vkCmdBindDescriptorSets
warning(zld): _vkCmdBindDescriptorSets
warning(zld): _vkBeginCommandBuffer
warning(zld): _vkBeginCommandBuffer
warning(zld): _vkAllocateMemory
warning(zld): _vkAllocateMemory
warning(zld): _vkAllocateDescriptorSets
warning(zld): _vkAllocateDescriptorSets
warning(zld): _vkAllocateCommandBuffers
warning(zld): _vkAllocateCommandBuffers
warning(zld): _vkAcquireNextImageKHR
warning(zld): _vkAcquireNextImageKHR
warning(zld): _vkResetCommandPool
warning(zld): _vkResetCommandPool
warning(zld): _vkUpdateDescriptorSets
warning(zld): _vkUpdateDescriptorSets
```

Comment (moosichu) [reply src/link/MachO/Zld.zig:1837]: Oh haha - I've also just realised that the if logic I intended was inverted - whoops.

Comment (moosichu) [reply src/link/MachO/Zld.zig:1836]: Yep - that explains why the built binary never actually ran properly (I only had the compile "successfully" complete) - no that I've actually got the if statement doing what I intended I do now have a functioning executable! :D 

However- now I need to actually solve the problem that is causing a need for this (because it should not be needed).

**Comment by moosichu** [comment]
OK - I've been able to build a little on feedback & identify at least one bug which has now been fixed - so the compiled executable (which I have been cross-compiling from my Linux desktop because I don't have a zig source-build available on my Mac) actually ran this time! 

So I believe the only _functional_ problem remaining is the fact that I've had to add that hacky if statement to Zld.zig - and I need to found out what is wrong with the previous logic to cause that. Then the rest of the work is just robustifying & cleaning-up what I have written and making it PR-ready, and we should be in a good place.

File: lib/std/io/reader.zig
Comment (moosichu) [reply lib/std/io/reader.zig:300]: Yeah - I'm not sure I'm super comfortable changing the prototype of readStruct, but I'm also not sure if adding this function is the right solution either.

File: src/link/MachO/Zld.zig
Comment (moosichu) [reply src/link/MachO/Zld.zig:1836]: OK - I'm done for the day, so will investigate this issue another time. If you have any ideas of what might be causing these duplicate symbols please do let me know. Basically - if I don't have this if statement - I get a panic in putNoClobber.

Comment (kubkon) [reply src/link/MachO/Zld.zig:1836]: I think you might wanna rebase onto `zld-tapi` locally, remove any tweaks to symbol resolutions, and see if it works fine then. Kinda hoping that changes in that branch made the symbol resolution logic more robust.

Comment (moosichu) [reply src/link/MachO/Zld.zig:1836]: Yes! That seems to resolve the issue. Just running the built executable on a Mac now to check that it works. 

If this is the thing that solves this then - what would you like me to do w.r.t this branch/PR relative to the `zld-tapi` branch?

Comment (moosichu) [reply src/link/MachO/Zld.zig]: The built executable ran successfully! :) I've pushed the removal of the ZLD changes to the branch here - but let me know what would be good for me to do regarding this.

Also if you have any further feedback on the current state of the PR that would be really helpful - I believe I've resolved some of it, but as this is my first real chunk of zig code that I've written (outside of a build.zig file for me C++ project), I'm aware it might be a bit idiosyncratic. 

If it does reach the point where from a code-POV it's an OK set of changes to merge, would you also like me to squash the commits?

File: lib/std/io/reader.zig
Comment (LemonBoy) [reply lib/std/io/reader.zig:300]: Given this change is orthogonal to the PR I'd take this off the patch and use `readStruct` + `bswapAllFields`.

File: lib/std/macho.zig
Comment (LemonBoy) [inline lib/std/macho.zig:32]: The formatting looks wonky, please run `zig fmt` over all the files you touch (that's a pre-requisite for the CI to pass).

File: lib/std/mem.zig
Comment (LemonBoy) [inline lib/std/mem.zig:1547]: `Elf32_Chdr` is not defined here, I'd replace this test with something like this:
```zig
const T = extern struct {
    f0: u8,
    f1: u16,
    f2: u32,
};
var s = T{
    .f0 = 0x12,
    .f1 = 0x1234,
    .f2 = 0x12345678,
};
bswapAllFields(T, &s);
try std.testing.expectEqual(T{
    .f0 = 0x12,
    .f1 = 0x3412,
    .f2 = 0x78563412,
}, s);
```

File: src/link/MachO/Dylib.zig
Comment (LemonBoy) [inline src/link/MachO/Dylib.zig:75]: If the function is not used outside of `Dylib` I'd drop the `pub`.

File: src/link/MachO/Dylib.zig
Comment (LemonBoy) [inline src/link/MachO/Dylib.zig:113]: Little tip, `while` expressions accept an `else` branch that's executed when the loop completed or, in other words, no `break` was executed. You can drop the `found_arch` flag using this nice construction.

File: src/link/MachO/Dylib.zig
Comment (LemonBoy) [inline src/link/MachO/Dylib.zig:93]: If you want to be extra-correct you could detect all the valid `FAT_*` numbers here and report an error if the endianness/format doesn't match the only format you're parsing here.

File: src/link/MachO/Dylib.zig
Comment (LemonBoy) [inline src/link/MachO/Dylib.zig:100]: RE: naming conventions. Variables are written in `snake_case`, methods in `camelCase`, types start with a capital letter and constants are (usually) in all-caps.

Comment (moosichu) [reply src/link/MachO/Dylib.zig:100]: Yes cheers! This was just out of habit from - will make sure to sort these!

File: src/link/MachO/Dylib.zig
Comment (moosichu) [reply src/link/MachO/Dylib.zig:93]: The fat headers are _always_ big endian. I don't know if it's worth checking for the incorrect endianess as that would imply something has gone horribly wrong with the source file?

Comment (LemonBoy) [reply src/link/MachO/Dylib.zig:93]: My bad, I missed that part in the `fat.h` header:
```
* All structures defined here are always written and read to/from disk
* in big-endian order.
```
Carry on then, and sorry for the noise!

File: src/link/MachO/Dylib.zig
Comment (moosichu) [reply src/link/MachO/Dylib.zig]: So I've done this work now - the only potential concern is that we are now reading the fat headers multiple times from disk (once to check if it embeds dylibs & once to search for a good candidate when parsing), I think it would be preferable to do this work only once (especially for people not compiling on an SSD).

File: src/link/MachO/Dylib.zig
Comment (moosichu) [reply src/link/MachO/Dylib.zig]: I've added a comment to the code though now to make this clearer :)

File: src/link/MachO/Dylib.zig
Comment (kubkon) [inline src/link/MachO/Dylib.zig:141]: Given that we may want to parse additional file contents in the future, I think it would make more sense to preserve the offset in `Dylib` and add it whenever we `pread` from the file.

File: src/link/MachO/Dylib.zig
Comment (kubkon) [inline src/link/MachO/Dylib.zig:80]: Could you move the `log.err` informing which arch was passed to here?

File: src/link/MachO/Dylib.zig
Comment (kubkon) [inline src/link/MachO/Dylib.zig:128]: Could you move this into `decodeArch`? Then, you can simplify it to `try decodeArch` without the need to switch on the error variants.

File: src/link/MachO/Dylib.zig
Comment (kubkon) [inline src/link/MachO/Dylib.zig:117]: I think you could simplify this greatly using labeled blocks like so:

```zig
const library_offset: i64 = offset: {
    const fat_header = try readFatStruct(self.file.?.reader(), macho.fat_header);
    if (fat_header.magic != macho.FAT_MAGIC) break :offset 0;

    var fat_arch_index: u32 = 0;
    while (fat_arch_index < fat_header.nfat_arch) : (fat_arch_index += 1) {
        const fat_arch = try readFatStruct(self.file.?.reader(), macho.fat_arch);
        // If we come across an architecture that we do not know how to handle, that's
        // fine because we can keep looking for one that might match.
        const lib_arch = decodeArch(fat_arch.cputype) catch |err| switch (err) {
            error.UnsupportedCpuArchitecture => continue,
            else => |e| return e,
         };
         if (lib_arch == self.arch.?) {
             // We have found a matching architecture!
             break :offset fat_arch.offset;
         }
     } else {
         log.err("Could not find matching cpu architecture in fat library: expected {s}", .{self.arch.?});
         return error.MismatchedCpuArchitecture;
     }
};
try self.file.?.seekTo(library_offset);
```

File: src/link/MachO/Dylib.zig
Comment (kubkon) [inline src/link/MachO/Dylib.zig:238]: I don't think this var is used anywhere?

**Review by kubkon — CHANGES_REQUESTED** [review]

File: src/link/MachO/Dylib.zig
Comment (moosichu) [reply src/link/MachO/Dylib.zig:128]: That is what it was originally - but I moved it here because a fat lib containing an unsupported arch itself to me doesn't feel like it should cause an error (because we only care about it supporting the target arch at all) . It's only an error on normal dylib.

Comment (kubkon) [reply src/link/MachO/Dylib.zig:128]: This will always be an error in zld since we most likely wont support anything else in the near future. Also, this helper should really be reused all over the MachO linker but that's an exercise for the future.

Comment (moosichu) [reply src/link/MachO/Dylib.zig:128]: Sorry, maybe I should be less cautious about this. But isn't it within the bounds of possibility that a certain library might only be available/distributed as a fat lib containing older architectures (like PowerPC) & x86 - we still want ZLD to link against the x86 part of that right without complaining about the fact that it contains another arch within it (this could also hypothetically extend to the future).

Maybe you are right though in that pragmatically - Apple are (well, Apple) and working with fat libs that contain any arch other than ARM64 & x86-64 is not going to happen for the foreseeable future (and fat lib support will become less relevant anyway as Apple Silicon becomes more prevalent) - and older fat libs that were created at previous transition points that can't be rebuilt from source are probably more than old enough to not warrant needing to support.

File: src/link/MachO/Dylib.zig
Comment (moosichu) [reply src/link/MachO/Dylib.zig:117]: Based on our other conversation this should be turned into a try right (and just error if we come across unsupported arch):
```zig

// If we come across an architecture that we do not know how to handle, that's
// fine because we can keep looking for one that might match.
const lib_arch = decodeArch(fat_arch.cputype) catch |err| switch (err) {
    error.UnsupportedCpuArchitecture => continue,
    else => |e| return e,
};
```

File: src/link/MachO/Dylib.zig
Comment (kubkon) [reply src/link/MachO/Dylib.zig:128]: Ah, I think I see what you mean: this would print a log message to the stderr when iterating over the archs contained within the fat lib which you are right, we do not want to do. If someone packs powerpc, x86, in addition to arm64 or x86_64, then that's fine as long as we find a lib we can link against. Nvm then, ignore my comment.

File: src/link/MachO/Dylib.zig
Comment (moosichu) [reply src/link/MachO/Dylib.zig]: Are you happy if I leave things as is currently with the PR, or should we address these concerns?

File: src/link/MachO/Dylib.zig
Comment (kubkon) [reply src/link/MachO/Dylib.zig:84]: The CI is failing since this is an used capture here. You capture any other potential architecture as `value` but never use it. In this case, just omit the capture altogether like so:

```zig
else => {
```

Comment (kubkon) [reply src/link/MachO/Dylib.zig:84]: The CI is failing since this is an used capture here. You capture any other potential architecture as `value` but never use it. In this case, just omit the capture altogether like so:

```zig
else => {
```

**Review by kubkon — CHANGES_REQUESTED** [review]

File: src/link/MachO/Zld.zig
Comment (moosichu) [reply src/link/MachO/Zld.zig]: I've squashed the commits (because it felt like thee right thing to do) - the only thing I'm still unsure about is if this ready to merge as-is (because the changes were only actually functional for me when merged with the zld-tapi branch). What is the eta for the zld-tapi branch to be merged into master?

Comment (kubkon) [reply src/link/MachO/Zld.zig]: `zld-tapi` landing in master should happen really soon.

**Comment by moosichu** [comment]
Beyond the multiple file-seek issue I mentioned up-top, I've addressed all the feedback (I think) so unless that is an issue I think this is now ready to be merged. Let me know if there are any other desirable changes that I should make!

**Comment by kubkon** [comment]
@moosichu #9171 has now been merged. Could you rebase onto master please?

**Comment by moosichu** [comment]
> @moosichu #9171 has now been merged. Could you rebase onto master please?

Sure thing! Will be able to do it later this evening.

**Comment by moosichu** [comment]
All done @kubkon ! :)

**Pull request merged**
